packages/ckeditor5-source-editing/src/sourceediting.js

```diff
@@ -0,0 +1,320 @@
+/**
+ * @license Copyright (c) 2003-2021, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module source-editing/sourceediting
+ */
+
+import { Plugin, PendingActions } from 'ckeditor5/src/core';
+import { ButtonView } from 'ckeditor5/src/ui';
+import { createElement, ElementReplacer } from 'ckeditor5/src/utils';
+
+import '../theme/sourceediting.css';
+
+// TODO: create icon
+// import sourceEditingIcon from '../theme/icons/sourceediting.svg';
+
+const COMMAND_FORCE_DISABLE_ID = 'SourceEditingMode';
+
+/**
+ * The source editing feature.
+ *
+ * It provides the possibility to view and edit the source of the document.
+ *
+ * For a detailed overview, check the {@glink features/source-editing source editing feature documentation} and the
+ * {@glink api/source-editing package page}.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class SourceEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'SourceEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ PendingActions ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Flag indicating whether the document source mode is active.
+		 *
+		 * @observable
+		 * @member {Boolean}
+		 */
+		this.set( 'isSourceEditingMode', false );
```

Comment:
As this is an observable property, do we need to maintain also the `sourceEditing` event? IMO we can handle all cases via changing this property and handling its value changes.

Reply:
Changed as proposed. Plugin now listen to the `change:isSourceEditingMode`.